File: 240/SearchA2DMatrixII.java
/**
 * SearchA2DMatrixII
 */
public class SearchA2DMatrixII {

    public boolean searchMatrix(int[][] matrix, int target) {
        return searchMatrixUtil(matrix, target, matrix.length - 1, 0);
    }

    public boolean searchMatrixUtil(int[][] matrix, int target, int x, int y) {
        if (matrix.length == 0 || matrix[0].length == 0)
            return false;

        int m = matrix.length;
        int n = matrix[0].length;

        if (x >= 0 && x < m && y >= 0 && y < n) {
            if (matrix[x][y] == target)
                return true;
            else if (matrix[x][y] > target)
                return searchMatrixUtil(matrix, target, x - 1, y);
            else
                return searchMatrixUtil(matrix, target, x, y + 1);
        } else
            return false;
    }

    public static void main(String[] args) {
        int[][] matrix = new int[][]{{-5}};
        int target = -5;
        System.out.println(new SearchA2DMatrixII().searchMatrix(matrix, target));
    }
}

File: 240/SearchA2DMatrixII.scala
object SearchA2DMatrixII {
  def searchMatrix(matrix: Array[Array[Int]], target: Int): Boolean = if (matrix.nonEmpty && matrix.head.nonEmpty) {
    val m = matrix.length
    val n = matrix.head.length

    @annotation.tailrec
    def search(x: Int, y: Int): Boolean = if (x >= 0 && x < m && y >= 0 && y < n) {
      if (matrix(x)(y) == target) true
      else if (matrix(x)(y) > target) search(x - 1, y)
      else search(x, y + 1)
    } else false

    search(m-1, 0)
  } else false

  def main(args: Array[String]): Unit = {
    val matrix = Array(Array(-5))
    val target = -5
    println(searchMatrix(matrix, target))
  }
}
